src/java/org/apache/cassandra/tcm/log/SystemKeyspaceStorage.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.cassandra.tcm.MetadataSnapshots;
+import org.apache.cassandra.tcm.Period;
+import org.apache.cassandra.tcm.Transformation;
+
+import static org.apache.cassandra.cql3.QueryProcessor.executeInternal;
+import static org.apache.cassandra.db.SystemKeyspace.LocalMetadataLog;
+
+/**
+ * {@code LogStorage} storing CMS log entries into the {@code system.local_metadata_log} table.
+ */
+public class SystemKeyspaceStorage implements LogStorage
+{
+    private static final Logger logger = LoggerFactory.getLogger(SystemKeyspaceStorage.class);
+    public static final String NAME =  org.apache.cassandra.db.SystemKeyspace.METADATA_LOG;
+
+    /**
+     * Generation is used as a timestamp for automatic table creation on startup.
+     * If you make any changes to the tables below, make sure to increment the
+     * generation and document your change here.
+     * <p>
+     * gen 0: original definition in 5.0
+     */
+    public static final long GENERATION = 0;
+
+    private final Supplier<MetadataSnapshots> snapshots;
+
+    public SystemKeyspaceStorage()
+    {
+        this(() -> ClusterMetadataService.instance().snapshotManager());
+    }
+
+    @VisibleForTesting
+    public SystemKeyspaceStorage(Supplier<MetadataSnapshots> snapshots)
+    {
+        this.snapshots = snapshots;
+    }
+
+    // This method is always called from a single thread, so doesn't have to be synchonised.
+    @Override
+    public void append(long period, Entry entry)
+    {
+        try
+        {
+            // TODO get lowest supported metadata version from ClusterMetadata
+            ByteBuffer serializedTransformation = entry.transform.kind().toVersionedBytes(entry.transform);
+            String query = String.format("INSERT INTO %s.%s (period, epoch, current_epoch, entry_id, transformation, kind) VALUES (?,?,?,?,?,?)",
+                                         SchemaConstants.SYSTEM_KEYSPACE_NAME, NAME);
+            executeInternal(query, period, entry.epoch.getEpoch(), entry.epoch.getEpoch(),
+                            entry.id.entryId, serializedTransformation, entry.transform.kind().toString());
+            // todo; should probably not flush every time, but it simplifies tests
+            Keyspace.open(SchemaConstants.SYSTEM_KEYSPACE_NAME).getColumnFamilyStore(NAME).forceBlockingFlush(ColumnFamilyStore.FlushReason.INTERNALLY_FORCED);
+        }
+        catch (Throwable t)
+        {
+            logger.error("Could not persist the entry {} proceeding with in-memory commit.", entry, t);
+        }
+    }
+
+    /**
+     * Checks if the metdata log table contains at least one row.
+     * @return {@code true} if the metdata log table contains at least one row, {@code false} otherwise.
+     */
+    public synchronized static boolean hasAnyEpoch()
+    {
+        String query = String.format("SELECT epoch FROM %s.%s LIMIT 1", SchemaConstants.SYSTEM_KEYSPACE_NAME, NAME);
+
+        for (UntypedResultSet.Row row : executeInternal(query))
```

Comment:
executeInternal can return null?

Reply:
I suppose it returns null when the statement does not return rows (for example, a schema update, or grant/revoke)

Reply:
We need to take precautions with this loop.